tests/lib/workflows/test_email_shutoff.py

```diff
@@ -0,0 +1,181 @@
+    Tests the case where a project id has been given
+    """
+
+    mock_project_id = NonCallableMock()
+
+    with patch('workflows.email_shutoff.ServerQuery') as mock_server_query:
+        query_shutoff_vms(mock_project_id)
+
+    server_object = mock_server_query.return_value
+
+    server_object.run.assert_called_once_with("openstack", as_admin=True, from_projects=mock_project_id)
+
+    user_query = server_object.then.return_value
+
+    user_query.run.assert_called_once_with("openstack")
+
+
+def test_query_server_get_user_details():
+    """
+    Tests that we get the details of the user from a server query
+    This tests the user query section of the query chaining used in the method
+    """
+    mock_project_id = None
+
+    with patch('workflows.email_shutoff.ServerQuery') as mock_server_query:
+        res = query_shutoff_vms(mock_project_id)
+
+    server_object = mock_server_query.return_value
+    server_object.then.assert_called_once_with("USER_QUERY", keep_previous_results=True)
+
+    user_query = server_object.then.return_value
+    user_query.select.assert_called_once_with(UserProperties.USER_EMAIL)
+    select_object = user_query.select.return_value
+
+    select_object.group_by.assert_called_once_with(UserProperties.USER_NAME)
+
+    user_query.run.assert_called_once()
+
+    user_query.to_string.assert_called_once()
+
+    assert res == user_query.to_string.return_value
+
+
+def test_prepare_user_server_email():
+    """
+    Tests that we extract the user_email and names of servers
+    from the server_query_result parameter
+    """
+    fake_server_a = {'server_name': NonCallableMock(), 'user_email': 'example@example.com'}
+    fake_server_b = {'server_name': NonCallableMock(), 'user_email': 'example@example.com'}
+
+    fake_query_result = [fake_server_a, fake_server_b]
+
+    expected_email = fake_server_a['user_email']
+    expected_server_names = [fake_server_a['server_name'], fake_server_b['server_name']]
+
+    result_email, result_server_names = prepare_user_server_email(fake_query_result)
+
+    assert result_email == expected_email
+    assert result_server_names == expected_server_names
+
+
+def test_send_user_email():
+    """
+    Test that an email is sent with the send_email method
+    """
+    mock_smtp_account = NonCallableMock()
+    mock_email = NonCallableMock()
+    mock_subject = NonCallableMock()
+    with patch('workflows.email_shutoff.Emailer') as mock_emailer:
+        send_user_email(mock_smtp_account, mock_email, mock_subject)
+
+    mock_emailer.assert_called_once()
+    email_object = mock_emailer.return_value
+
+    expected_email_params = EmailParams(
+        subject=mock_subject,
+        email_from='cloud-support@stfc.ac.uk',
+        email_to=[mock_email],
+        email_templates=[]
+    )
+
+    email_object.send_emails.assert_called_once_with([expected_email_params])
+
+
+@patch('workflows.email_shutoff.Emailer')
```

Comment:
Nit this one uses an `@patch` but the one above uses a `with` for the same code path, could we change this to use a `with` context manager like the last one?